build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/SplitPackagesAuditTask.java

```diff
@@ -0,0 +1,324 @@
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to create marker file", e);
+            }
+        }
+
+        private Map<String, List<File>> getDependencyPackages() {
+            Map<String, List<File>> packages = new HashMap<>();
+            for (File classpathElement : getParameters().getClasspath().getFiles()) {
+                if (classpathElement.exists() == false) {
+                    continue;
+                }
+                for (String packageName : readPackages(classpathElement)) {
+                    packages.computeIfAbsent(packageName, k -> new ArrayList<>()).add(classpathElement);
+                }
+            }
+            if (LOGGER.isInfoEnabled()) {
+                List<String> msg = new ArrayList<>();
+                msg.add("Packages from dependencies:");
+                packages.entrySet().stream().sorted().forEach(e -> msg.add("  -" + e.getKey() + " -> " + e.getValue()));
+                LOGGER.info(String.join(System.lineSeparator(), msg));
+            }
+            return packages;
+        }
+
+        private Map<String, Set<String>> findSplitPackages(Set<String> dependencyPackages) {
+            Map<String, Set<String>> splitPackages = new HashMap<>();
+            for (File srcDir : getParameters().getSrcDirs().get()) {
+                try {
+                    walkJavaFiles(srcDir.toPath(), ".java", classfile -> {
+                        String packageName = getPackageName(classfile);
+                        String className = classfile.subpath(classfile.getNameCount() - 1, classfile.getNameCount()).toString();
+                        className = className.substring(0, className.length() - ".java".length());
+                        className = className.replace('$', '.'); // for inner class names
+                        LOGGER.info("Inspecting " + classfile + System.lineSeparator()
+                            + "  package: " + packageName + System.lineSeparator()
+                            + "  class: " + className);
+                        if (dependencyPackages.contains(packageName)) {
+                            splitPackages.computeIfAbsent(packageName, k -> new HashSet<>()).add(packageName + "." + className);
+                        }
+                    });
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+            if (LOGGER.isInfoEnabled()) {
+                List<String> msg = new ArrayList<>();
+                msg.add("Split packages:");
+                splitPackages.entrySet().stream().sorted().forEach(e -> msg.add("  -" + e.getKey() + " -> " + e.getValue()));
+                LOGGER.info(String.join(System.lineSeparator(), msg));
+            }
+            return splitPackages;
+        }
+
+        private void filterSplitPackages(Map<String, Set<String>> splitPackages) {
```

Comment:
Goning back and mutating our findings seems unnecessarily complicated. Why not just filter these out directly in `splitPackages()`. Since we already have the package/classname there, we can just check against the filter list before adding to the result set.

Reply:
I prefer the separation, since the algorithms are clearer (how to find the split packages vs reading through filters).